src/metaswitch/clearwater/cluster_manager/etcd_synchronizer.py

```diff
@@ -32,121 +32,88 @@
+            if etcd_value is not None:
+                _log.debug("Got new state %s from etcd" % etcd_value)
+                cluster_info = ClusterInfo(etcd_value)
+
+                # This node can only leave the cluster if the cluster is in a
+                # stable state. Check the leaving flag and the cluster state. If
+                # necessary, set this node to WAITING_TO_LEAVE. Otherwise, kick
+                # the FSM.
+                if (self._leaving_flag and
+                        cluster_info.can_leave(allow_errors=self.force_leave)):
                     _log.info("Cluster is in a stable state, so leaving the cluster now")
                     new_state = constants.WAITING_TO_LEAVE
                 else:
-                    local_state = self.calculate_local_state(cluster_view)
-                    new_state = self._fsm.next(local_state,
-                                               cluster_state,
-                                               cluster_view)
+                    new_state = self._fsm.next(cluster_info.local_state(self._ip),
+                                               cluster_info.cluster_state,
+                                               cluster_info.view)
 
                 # If we have a new state, try and write it to etcd.
                 if new_state is not None:
-                    self.write_to_etcd(cluster_view, new_state)
+                    self.write_to_etcd(cluster_info, new_state)
                 else:
                     _log.debug("No state change")
             else:
-                _log.warning("read_from_etcd returned None, indicating a failure to get data from etcd")
+                _log.warning("read_from_etcd returned None, " +
+                             "indicating a failure to get data from etcd")
 
         _log.info("Quitting FSM")
         self._fsm.quit()
-        self.executor.shutdown(wait=False)
-
-    def parse_cluster_view(self, view):
-        try:
-            return json.loads(view)
-        except:
-            return {}
 
     # This node has been asked to leave the cluster. Check if the cluster is in
     # a stable state, in which case we can leave. Otherwise, set a flag and
     # leave at the next available opportunity.
     def leave_cluster(self):
-        _log.info("Trying to leave the cluster - plugin %s" % self._plugin.__class__.__name__)
+        _log.info("Trying to leave the cluster - plugin %s" %
+                  self._plugin.__class__.__name__)
+
         if not self._plugin.should_be_in_cluster():
             _log.info("No need to leave remote cluster - just exit")
             self._terminate_flag = True
             return
-        result = self._client.read(self._key, quorum=True)
-        cluster_view = self.parse_cluster_view(result.value)
-        self._index = result.modifiedIndex
 
-        cluster_state = self.calculate_cluster_state(cluster_view)
+        etcd_result, idx = self.read_from_etcd(wait=False)
```

Comment:
All reads are quorum reads (and all waits are non-quorum), so I don't think a quorum parameter adds anything.

On `update_from_etcd` vs. `read_from_etcd`, because `update_from_etcd` changes internal state, I thought it was cleaner if only one thread calls it - otherwise they might race. I'll comment the method descriptions.